scanpy/neighbors/__init__.py

```diff
@@ -173,6 +173,85 @@ def neighbors_update(adata, adata_new, k=10, queue_size=5, random_state=0):
     return indices[:, :k], dists[:, :k]
 
 
+# Generates a timestamp for use in logging messages when verbose=True
+def ts():
+    import time
+    return time.ctime(time.time())
+
+
+def nearest_neighbors(
+    X, n_neighbors, metric, metric_kwds, angular, random_state, verbose=False
+):
+    """Compute the ``n_neighbors`` nearest points for each data point in ``X``
+    under ``metric``. This may be exact, but more likely is approximated via
+    nearest neighbor descent.
+
+    Parameters
+    ----------
+    X: array of shape (n_samples, n_features)
+        The input data to compute the k-neighbor graph of.
+
+    n_neighbors: int
+        The number of nearest neighbors to compute for each sample in ``X``.
+
+    metric: string or callable
+        The metric to use for the computation.
+
+    metric_kwds: dict
+        Any arguments to pass to the metric computation function.
+
+    angular: bool
+        Whether to use angular rp trees in NN approximation.
+
+    random_state: np.random state
+        The random state to use for approximate NN computations.
+
+    verbose: bool
+        Whether to print status data during the computation.
+
+    Returns
+    -------
+    knn_indices: array of shape (n_samples, n_neighbors)
+        The indices on the ``n_neighbors`` closest points in the dataset.
+
+    knn_dists: array of shape (n_samples, n_neighbors)
+        The distances to the ``n_neighbors`` closest points in the dataset.
+    """
+    if verbose:
+        print(ts(), "Finding Nearest Neighbors")
+
+    if metric == "precomputed":
+        # Note that this does not support sparse distance matrices yet ...
+        # Compute indices of n nearest neighbors
+        knn_indices = np.argsort(X)[:, :n_neighbors]
+        # Compute the nearest neighbor distances
+        #   (equivalent to np.sort(X)[:,:n_neighbors])
+        knn_dists = X[np.arange(X.shape[0])[:, None], knn_indices].copy()
+
+        rp_forest = []
+    else:
+        n_trees = 5 + int(round((X.shape[0]) ** 0.5 / 20.0))
+        n_iters = max(5, int(round(np.log2(X.shape[0]))))
+
+        from pynndescent import NNDescent
+        nn_descent = NNDescent(
+            X,
+            n_neighbors=n_neighbors,
+            metric=metric,
+            metric_kwds=metric_kwds,
+            random_state=random_state,
+            n_trees=n_trees,
+            n_iters=n_iters,
+            max_candidates=60,
+            verbose=verbose,
+        )
+        knn_indices, knn_dists = nn_descent._neighbor_graph
+        rp_forest = nn_descent._rp_forest
+
+    if verbose:
+        print(ts(), "Finished Nearest Neighbor Search")
+    return knn_indices, knn_dists, rp_forest
```

Comment:
Why don't we return `nn_descent` object here instead of the rp_forest? If we store it in `adata.uns` then we can build other methods for querying the graph. What do you think?

Reply:
Hm, but we need to be able to write it on disk and read it back.

Reply:
Yes, this whole discussion circles around "how we efficiently store things on disk while providing users convenient objects". The `Ingest` class will take care of several aspects relating to the graph (querying the graph for ingesting new data [we can also think about another name here]). You are right that there might be other use cases for querying the graph. For now, I'd remain conservative and simply keep the forest here. It's not even entering AnnData by default...

One could even think about infrastructure within AnnData. But let's have this discussion somewhere else.